File: Java/homework-5/src/com/testts/AbstractFileSystemNode.java
package com.testts;

import java.util.ArrayList;
import java.util.Collections;
import java.util.stream.Collectors;

public class AbstractFileSystemNode implements FileSystemNode{
    FileSystemNode parent = null;
    String name = null;

    @Override
    public final FileSystemNode getParent() {
        return parent;
    };

    @Override
    public String getName() {
        return name;
    }

    @Override
    public final String getPath() {
        FileSystemNode ancestor = this;
        ArrayList<String> folders = new ArrayList<String>();
        do {
            ancestor = ancestor.getParent();
            folders.add(ancestor.getName());
        } while (ancestor.getParent() != null);
        Collections.reverse(folders);
        return folders.stream().collect(
                Collectors.joining("/"));
    }

    @Override
    public void setParent(FileSystemNode newParent) {
        this.parent = newParent;
    }

    @Override
    public String getExtension() {
        String[] filenameParts = this.name.split("\\.");
        return filenameParts.length > 1 ? filenameParts[1] : filenameParts[0];
    }
}

File: FileSystemTutorial/src/main/kotlin/AbstractFileSystemNode.kt

import java.util.stream.Collectors

import java.util.Collections

import java.util.ArrayList


abstract class AbstractFileSystemNode() : FileSystemNode {
    var parent: FileSystemNode? = null
    var name: String? = null

    override fun getNodeParent(): FileSystemNode? {
        return parent
    }

    override fun getNodeName(): String? {
        return name
    }

    override fun getNodePath(): String? {
        var ancestor: FileSystemNode? = this
        val folders = ArrayList<String?>()
        do {
            ancestor = ancestor!!.getNodeParent()
            folders.add(ancestor!!.getNodeName())
        } while (ancestor!!.getNodeParent() != null)
        Collections.reverse(folders)
        return folders.stream().collect(
            Collectors.joining("/")
        )
    }

    override fun setNodeParent(newParent: FileSystemNode?) {
        parent = newParent
    }

    override fun getNodeExtension(): String? {
        val filenameParts = name!!.split("\\.").toTypedArray()
        return if (filenameParts.size > 1) filenameParts[1] else filenameParts[0]
    }
}